### DISCUSSION TITLE
Creating a table in a rpc function doesn't grant needed permissions

### DISCUSSION THREAD

**Discussion by dancamdev** [supabase/supabase#10224] [Questions]
I'm currently using this function to create a new table in the database on a monthly basis.

```sql
create or replace function create_monthly_stats_table() returns void as $$
  begin
    -- create a table if it doesn't exist
    -- table name is the current month
    -- table columns uid reference locali table, viewed, accepted, map, call, web
    execute 'create table if not exists ' || to_char(current_date, 'stats_YYYY_MM') || ' (id int8 generated by default as identity primary key, uid uuid references locali(uid), viewed numeric, accepted numeric, map numeric, call numeric, web numeric)';

    -- set table schema to public
    execute 'alter table ' || to_char(current_date, 'stats_YYYY_MM') || ' set schema public';

    -- grant select, insert, update on table any role
    execute 'grant select, insert, update on ' || to_char(current_date, 'stats_YYYY_MM') || ' to public';
  end;
$$ language plpgsql;
```

It works great if I invoke it using the SQL Editor within the supabase dashboard.

But whenever I try to call it from the client

```JavaScript
const {error} = await supabase.rpc('create_monthly_stats_table');
```

I get the following error:

```Bash
 Error {"code": "42501", "details": null, "hint": null, "message": "permission denied for schema public"}
```

**Comment by GaryAustin1** [comment]
What user is calling this function?  The function is going to run with their permissions.  You could make it a "security definer" function and it will run with the creator's permissions.

BUT this seems very dangerous as anyone can run rpc functions unless you secure them somehow with grants or at least checks in the function. 

Creating tables on the fly is not normally recommend.

**Reply by dancamdev** [reply]
It's getting called by an unauthenticated user - to my defense, this function is actually getting called by another function though.

Still, taking a step back, is there a better way to create a monthly stats table automatically? Am I missing something?
